java/arcs/core/data/HandleMode.kt

```diff
@@ -18,14 +18,26 @@ enum class HandleMode {
     Read,
     /** [Handle] is write only. */
     Write,
+    /** [Handle] is query only. */
+    Query,
     /** [Handle] is read-write. */
-    ReadWrite;
+    ReadWrite,
+    /** [Handle] is read-query. */
+    ReadQuery,
+    /** [Handle] is query-write. */
+    WriteQuery,
+    /** [Handle] is read-query-write. */
+    ReadWriteQuery;
 
     /** True if reading is supported by this mode. */
     val canRead: Boolean
-        get() = this != Write
+        get() = this == Read || this == ReadWrite || this == ReadQuery || this == ReadWriteQuery
 
     /** True if writing is supported by this mode. */
     val canWrite: Boolean
-        get() = this != Read
+        get() = this == Write || this == ReadWrite || this == WriteQuery || this == ReadWriteQuery
+
+    /** True if quering is supported by this mode. */
+    val canQuery: Boolean
+        get() = this == Query || this == ReadQuery || this == WriteQuery || this == ReadWriteQuery
```

Comment:
Rather than adding a bunch of boolean logic here, it could be made declarative by adding parameters to the enum's constructor.

```kotlin
enum class HandleMode(
    val canRead: Boolean = false, 
    val canWrite: Boolean = false, 
    val canQuery: Boolean = false
) {
    Read(canRead = true),
    Write(canWrite = true),
    Query(canQuery = true),
    ReadWrite(canRead = true, canWrite = true),
    ReadQuery(canRead = true, canQuery = true),
    WriteQuery(canWrite = true, canQuery = true),
    ReadWriteQuery(canRead = true, canWrite = true, canQuery = true);
}
```

Reply:
Can we use delegated constructors on enum constructors? It would be great the be able to say
`ReadWrite: Read, Write` or something similar.

Reply:
I think if you use sealed classes instead of enums, you can do whatever you can do with regular classes, but it probably won't save any keystrokes. If these were val's, you could use an infix operator to compose stuff, e.g. 

```
class HandleMode(...) {
   companion object {
     val Read = HandleMode(canRead = true)
     val Write = HandleMode(canWrite = true)
     val ReadWrite = Read + Write
   }
   operator plus(mode: HandleMode) = HandleMode(
      canRead = this.canRead || mode.canRead, 
      canWrite = this.canWrite || mode.canWrite
  )
}
```

Reply:
You could go old-school and use bit-masks. :D 

```kotlin
object HandleMode {
    const val Read = 0b001
    const val Write = 0b010
    const val Query = 0b100
    const val ReadWrite = Read or Write
    const val ReadQuery = Read or Query
    const val WriteQuery = Write or Query
    const val ReadWriteQuery = Read or Write or Query
}
```